types/prop-types/index.d.ts

```diff
@@ -1,33 +1,49 @@
 // Type definitions for prop-types 15.5
 // Project: https://github.com/reactjs/prop-types
 // Definitions by: DovydasNavickas <https://github.com/DovydasNavickas>
+//                 Ferdy Budhidharma <https://github.com/ferdaber>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.2
+// TypeScript Version: 2.8
 
-export type Validator<T> = (object: T, key: string, componentName: string, ...rest: any[]) => Error | null;
+import { ReactNode, ReactElement } from 'react';
 
-export interface Requireable<T> extends Validator<T> {
-    isRequired: Validator<T>;
+export type IsOptional<T> = undefined | null extends T ? true : undefined extends T ? true : null extends T ? true : false;
+
+export type RequiredKeys<V> = { [K in keyof V]: V[K] extends Validator<infer T> ? IsOptional<T> extends true ? never : K : never }[keyof V];
+export type OptionalKeys<V> = Exclude<keyof V, RequiredKeys<V>>;
+export type InferPropsInner<V> = { [K in keyof V]: InferType<V[K]>; };
+
+export type Validator<T> = (props: object, propName: string, componentName: string, location: string, propFullName: string, _NOMINAL_TYPE_HACK_DO_NOT_USE?: T) => Error | null;
```

Comment:
`_NOMINAL_TYPE_HACK_DO_NOT_USE`? I'd rather see `Validator` as an interface and usage of a unique symbol named property for the same purpose.

Especially since function argument type variance is checked inconsistently (strict or no) :D

Is there a reason it's jammed into a parameter slot here?

Reply:
This was kind of my progression of attempts:
- Used a required unique symbol property with type T. Custom validator functions break because now they need to supply the unique symbol.
- Used an optional unique symbol property with type T. Custom validator functions are OK, but `oneOf` breaks because the assignment check is the opposite (antivariant?) direction (
- Used an optional secret parameter which is a bit more lax? (really just unsure how to word this, my category theory knowledge is so rudimentary)

I do recognize it looks odd, so I think it'd be fine to use an optional unique symbol property, just that `oneOf` will need to almost always need to have its generic parameter specified in the case of a union of literal types.